src/vanilla/store2.ts

```diff
@@ -0,0 +1,724 @@
+  }
+
+const isContinuablePromise = (
+  promise: unknown,
+): promise is ContinuablePromise<AnyValue> =>
+  typeof promise === 'object' && promise !== null && CONTINUE_PROMISE in promise
+
+const continuablePromiseMap: WeakMap<
+  PromiseLike<AnyValue>,
+  ContinuablePromise<AnyValue>
+> = new WeakMap()
+
+/**
+ * Create a continuable promise from a regular promise.
+ */
+const createContinuablePromise = <T>(
+  promise: PromiseLike<T>,
+  abort: () => void,
+  complete: () => void,
+): ContinuablePromise<T> => {
+  if (!continuablePromiseMap.has(promise)) {
+    let continuePromise: ContinuePromise<T>
+    const p: any = new Promise((resolve, reject) => {
+      let curr = promise
+      const onFullfilled = (me: PromiseLike<T>) => (v: T) => {
+        if (curr === me) {
+          p.status = FULFILLED
+          p.value = v
+          resolve(v)
+          complete()
+        }
+      }
+      const onRejected = (me: PromiseLike<T>) => (e: AnyError) => {
+        if (curr === me) {
+          p.status = REJECTED
+          p.reason = e
+          reject(e)
+          complete()
+        }
+      }
+      promise.then(onFullfilled(promise), onRejected(promise))
+      continuePromise = (nextPromise, nextAbort) => {
+        if (nextPromise) {
+          continuablePromiseMap.set(nextPromise, p)
+          curr = nextPromise
+          nextPromise.then(onFullfilled(nextPromise), onRejected(nextPromise))
+        }
+        abort()
+        abort = nextAbort
+      }
+    })
+    p.status = PENDING
+    p[CONTINUE_PROMISE] = continuePromise!
+    continuablePromiseMap.set(promise, p)
+  }
+  return continuablePromiseMap.get(promise) as ContinuablePromise<T>
+}
+
+const isPromiseLike = (x: unknown): x is PromiseLike<unknown> =>
+  typeof (x as any)?.then === 'function'
+
+const getPendingContinuablePromise = (atomState: AtomState) => {
+  const value: unknown = (atomState as any).s?.v
+  if (isContinuablePromise(value) && value.status === PENDING) {
+    return value
+  }
+  return null
+}
+
+/**
+ * State tracked for mounted atoms. An atom is considered "mounted" if it has a
+ * subscriber, or is a transitive dependency of another atom that has a
+ * subscriber.
+ *
+ * The mounted state of an atom is freed once it is no longer mounted.
+ */
+type Mounted = {
+  /** Set of listeners to notify when the atom value changes. */
+  readonly l: Set<() => void>
+  /** Set of mounted atoms that the atom depends on. */
+  readonly d: Set<AnyAtom>
+  /** Function to run when the atom is unmounted. */
+  u?: OnUnmount
+}
+
+/**
+ * Mutable atom state,
+ * tracked for both mounted and unmounted atoms in a store.
+ */
+type AtomState<Value = AnyValue> = {
+  /**
+   * Map of atoms that the atom depends on.
+   * The map value is value/error of the dependency.
+   */
+  readonly d: Map<AnyAtom, { readonly v: AnyValue } | { readonly e: AnyError }>
+  /** Set of atoms that depends on the atom. */
+  readonly t: Set<AnyAtom>
```

Comment:
As this holds a strong reference, it may cause memory leaks.
We should create a failing test case and fix it. (Probably by moving it to `Mounted`, or possibly with WeakRef but unlikely.)

Reply:
#2477